python/hail/typecheck/check.py

```diff
@@ -0,0 +1,202 @@
+        self.ec = element_checker
+        super(CollectionChecker, self).__init__()
+
+    def check(self, x):
+        passes = self.cc.check(x)
+        if passes:
+            for elt in x:
+                passes = passes and self.ec.check(elt)
+        return passes
+
+    def expects(self):
+        return '%s[%s]' % (self.cc.expects(), self.ec.expects())
+
+
+class DictChecker(TypeChecker):
+    def __init__(self, key_checker, value_checker):
+        self.kc = key_checker
+        self.vc = value_checker
+        super(DictChecker, self).__init__()
+
+    def check(self, x):
+        passes = isinstance(x, dict)
+        if passes:
+            for k, v in x.iteritems():
+                passes = passes and self.kc.check(k)
+                passes = passes and self.vc.check(v)
+        return passes
+
+    def expects(self):
+        return 'dict[%s, %s]' % (self.kc.expects(), self.vc.expects())
+
+
+def only(t):
+    if isinstance(t, type) or type(t) is ClassType:
+        return LambdaChecker(lambda x: isinstance(x, t), extract(t))
+    elif isinstance(t, TypeChecker):
+        return t
+    else:
+        raise RuntimeError("invalid typecheck signature: expected 'type' or 'TypeChecker', found '%s'" % type(t))
+
+
+def oneof(*args):
+    return MultipleTypeChecker([only(x) for x in args])
+
+
+def nullable(t):
+    return oneof(t, NoneType)
+
+
+def listof(t):
+    return CollectionChecker(only(list), only(t))
+
+
+def dictof(k, v):
+    return DictChecker(only(k), only(v))
+
+
+none = only(NoneType)
+
+anytype = LambdaChecker(lambda x: True, 'Any')
+
+strlike = oneof(str, unicode)
+
+integral = oneof(int, long)
+
+numeric = oneof(int, long, float)
+
+
+def typecheck(*targs, **kw):
+    if len(targs) != 0 and len(kw) != 0:
+        raise RuntimeError('cannot mix positional and keyword type checking. Name all typecheck arguments or none.')
+
+    use_kw = len(kw) != 0
+
+    def get_type(name, index):
+        if use_kw:
+            return kw[name]
+        else:
+            return targs[index]
+
+    def _typecheck(f, *args, **kwargs):
+        argspec = getargspec(f)
+        argspec_args = argspec.args
+
+        self_cls = None
+
+        # strip the 'self' parameter, since this does not need to be checked
+        if len(argspec_args) > 0 and argspec_args[0] == 'self':
+            # try to get more information about the class to write better error messages
+            try:
+                self_cls = args[0].__class__
+            except:
+                pass
+            check_args = args[1:]
+            check_spec = argspec_args[1:]
+        else:
+            check_args = args
+            check_spec = argspec_args
+
+        # name of function, preceded by the class name if the _f_ is an instance method
+        name = '%s%s()' % (extract(self_cls).split('.')[-1] + '.' if self_cls else '', f.__name__)
+
+        # ensure that the typecheck signature is appropriate and matches the function signature
```

Comment:
I still think you should be able to typecheck args and kwargs as normal variables.